packages/substrate/src/fungible.ts

```diff
@@ -0,0 +1,158 @@
+} from '@buildwithsygma/core';
+import { Config, FeeHandlerType, ResourceType } from '@buildwithsygma/core';
+import type { ApiPromise, SubmittableResult } from '@polkadot/api';
+import type { SubmittableExtrinsic } from '@polkadot/api-base/types';
+import { BN } from '@polkadot/util';
+
+import { BaseTransfer } from './base-transfer.js';
+import type { Fungible, SubstrateFee, Transfer } from './types.js';
+import { deposit, getBasicFee, getFeeHandler, getPercentageFee, getLiquidity } from './utils';
+
+export type SubstrateAssetTransferRequest = {
+  sourceDomain: Domainlike;
+  destinationDomain: Domainlike;
+  sourceNetworkProvider: ApiPromise;
+  resource: string | SubstrateResource;
+  amount: bigint;
+  destinationAddress: string;
+  environment: Environment;
+};
+
+async function checkDestinationFungibleHandler(
+  destinationDomain: SubstrateConfig,
+  resourceId: string,
+  amount: bigint,
+  sourceNetworkProvider: ApiPromise,
+): Promise<void> {
+  const handlerAddress = destinationDomain.handlers.find(
+    h => h.type === ResourceType.FUNGIBLE,
+  )?.address;
+  if (!handlerAddress) {
+    throw new Error('No Fungible handler configured on destination domain');
+  }
+
+  const destinationResource = destinationDomain.resources.find(r => r.resourceId === resourceId);
+  const destinationHandlerBalance = await getLiquidity(
+    sourceNetworkProvider,
+    destinationResource as SubstrateResource,
+    handlerAddress,
+  );
+
+  if (destinationHandlerBalance < amount) {
+    throw new LiquidityError(destinationHandlerBalance);
+  }
+}
+
+export async function createSubstrateFungibleAssetTransfer(
+  transferRequestParams: SubstrateAssetTransferRequest,
+): Promise<SubstrateFungibleAssetTransfer> {
+  const config = new Config();
+  await config.init(transferRequestParams.environment);
+
+  const transfer = new SubstrateFungibleAssetTransfer(transferRequestParams, config);
+
+  const destinationDomain = config.getDomainConfig(transfer.destinationAddress) as SubstrateConfig;
+  await checkDestinationFungibleHandler(
+    destinationDomain,
+    transfer.resource.resourceId,
+    transfer.amount,
+    transfer.sourceNetworkProvider,
+  );
+
+  if (!(await transfer.isValidTransfer()))
+    throw new Error('Handler not registered, please check if this is a valid bridge route.');
+
+  return transfer;
+}
+
+/**
+ * @dev User should not instance this directly. All the (async) checks should be done in `createSubstrateFungibleAssetTransfer`
+ */
+export class SubstrateFungibleAssetTransfer extends BaseTransfer {
+  amount: bigint;
+  destinationAddress: string;
+
+  constructor(transfer: SubstrateAssetTransferRequest, config: Config) {
+    super(transfer, config);
+    this.sourceDomain = config.getDomain(transfer.sourceDomain);
+    this.destinationDomain = config.getDomain(transfer.destinationDomain);
+
+    this.amount = transfer.amount;
+    this.destinationAddress = transfer.destinationAddress;
+  }
+
+  setAmount(amount: bigint): this {
```

Comment:
is the return of `this` neccesary in this context to set the amount? AFAIK returning `this` allows you to do chaining, but not sure if `this` is intended.